File: src/lib/analysis.ts
// ── 通用分析引擎：相关性 / 事件研究 / 统计工具 ──
// 零依赖，服务端纯函数，供各决策模块复用

export interface SeriesPoint {
  date: string
  value: number
}

// ==== 统计基础 ====

export function mean(nums: number[]): number {
  if (nums.length === 0) return 0
  return nums.reduce((s, v) => s + v, 0) / nums.length
}

export function std(nums: number[]): number {
  if (nums.length < 2) return 0
  const m = mean(nums)
  return Math.sqrt(nums.reduce((s, v) => s + (v - m) ** 2, 0) / nums.length)
}

export function percentileRank(values: number[], v: number): number {
  if (values.length < 2) return 50
  const sorted = [...values].sort((a, b) => a - b)
  const below = sorted.filter(x => x < v).length
  return +((below / (sorted.length - 1)) * 100).toFixed(1)
}

export function zScore(values: number[], v: number): number {
  const m = mean(values)
  const s = std(values)
  return s === 0 ? 0 : (v - m) / s
}

export function quantile(sorted: number[], q: number): number {
  if (sorted.length === 0) return 0
  const pos = (sorted.length - 1) * q
  const lo = Math.floor(pos)
  const hi = Math.ceil(pos)
  if (lo === hi) return sorted[lo]
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo)
}

// ==== 收益率 ====

export function logReturns(points: SeriesPoint[]): SeriesPoint[] {
  const out: SeriesPoint[] = []
  for (let i = 1; i < points.length; i++) {
    const prev = points[i - 1].value
    const cur = points[i].value
    if (prev > 0 && cur > 0) {
      out.push({ date: points[i].date, value: Math.log(cur / prev) })
    }
  }
  return out
}

// ==== 按日期对齐两条序列（交集）====

export function alignByDate(a: SeriesPoint[], b: SeriesPoint[]): { date: string; a: number; b: number }[] {
  const bMap = new Map(b.map(p => [p.date, p.value]))
  const out: { date: string; a: number; b: number }[] = []
  for (const p of a) {
    const bv = bMap.get(p.date)
    if (bv != null && isFinite(bv) && isFinite(p.value)) {
      out.push({ date: p.date, a: p.value, b: bv })
    }
  }
  return out
}

// ==== 相关 ====

export function corr(ax: number[], bx: number[]): number {
  const n = Math.min(ax.length, bx.length)
  if (n < 3) return 0
  const a = ax.slice(-n)
  const b = bx.slice(-n)
  const ma = mean(a)
  const mb = mean(b)
  let cov = 0, sa2 = 0, sb2 = 0
  for (let i = 0; i < n; i++) {
    const da = a[i] - ma
    const db = b[i] - mb
    cov += da * db
    sa2 += da * da
    sb2 += db * db
  }
  if (sa2 === 0 || sb2 === 0) return 0
  return +(cov / Math.sqrt(sa2 * sb2)).toFixed(4)
}

// 滚动相关系数序列（窗口 N 个数据点）
export function rollingCorr(a: SeriesPoint[], b: SeriesPoint[], window: number): { date: string; value: number }[] {
  const aligned = alignByDate(a, b)
  const out: { date: string; value: number }[] = []
  for (let i = window; i <= aligned.length; i++) {
    const seg = aligned.slice(i - window, i)
    const r = corr(seg.map(p => p.a), seg.map(p => p.b))
    out.push({ date: seg[seg.length - 1].date, value: r })
  }
  return out
}

// 领先滞后相关：lag>0 表示 a 领先 b（b 的第 i+lag 日与 a 的第 i 日对齐）
// 适用于同一交易日历的两条序列（如 A 股资金流与沪深300）
export function leadLagCorr(a: SeriesPoint[], b: SeriesPoint[], maxLag: number): { lag: number; corr: number }[] {
  const bMap = new Map(b.map(p => [String(p.date), p.value]))
  const aDates = a.map(p => String(p.date))
  const out: { lag: number; corr: number }[] = []
  for (let lag = -maxLag; lag <= maxLag; lag++) {
    const ax: number[] = []
    const bx: number[] = []
    for (let i = 0; i < aDates.length; i++) {
      const j = i + lag
      if (j < 0 || j >= aDates.length) continue
      const bv = bMap.get(aDates[j])
      if (bv != null) {
        ax.push(a[i].value)
        bx.push(bv)
      }
    }
    out.push({ lag, corr: corr(ax, bx) })
  }
  return out
}

// ==== 事件研究 ====
// 给定事件日期流与资产价格序列，统计事件发生后 N 个交易日的收益分布
// 收益定义：以事件日之前最近一个交易日收盘价为基准（避免事件当天 BUG），
// 或事件日当天的收盘（若事件日有价格）→ 使用 eventDay 当收盘

export interface EventStudyRow {
  date: string
  rets: Record<string, number | null>
}

export interface HorizonStats {
  n: number
  mean: number
  median: number
  winRate: number
  p25: number
  p75: number
  best: number
  worst: number
}

export interface EventStudyResult {
  nEvents: number
  events: EventStudyRow[]
  horizons: Record<string, HorizonStats>
}

function forwardReturns(priceDates: string[], priceMap: Map<string, number>, baseDate: string, horizons: number[]): Record<string, number | null> {
  // 找到 baseDate 之前最近有价格的一天作为基准
  let baseIdx = -1
  for (let i = 0; i < priceDates.length; i++) {
    if (priceDates[i] <= baseDate) baseIdx = i
    else break
  }
  if (baseIdx < 0) return {}
  const base = priceMap.get(priceDates[baseIdx])
  if (base == null || base === 0) return {}
  const rets: Record<string, number | null> = {}
  for (const h of horizons) {
    const j = baseIdx + h
    if (j < priceDates.length) {
      const fwd = priceMap.get(priceDates[j])
      rets[String(h)] = fwd != null && fwd > 0 ? +((fwd / base) - 1).toFixed(4) : null
    } else {
      rets[String(h)] = null
    }
  }
  return rets
}

export function eventStudy(
  prices: SeriesPoint[],
  eventDates: string[],
  horizons: number[]
): EventStudyResult {
  const priceDates = prices.map(p => String(p.date)).sort()
  const priceMap = new Map(priceDates.map((d, i) => [d, prices[i].value]))

  const events: EventStudyRow[] = []
  for (const ev of eventDates) {
    const rets = forwardReturns(priceDates, priceMap, String(ev), horizons)
    events.push({ date: String(ev), rets })
  }

  const horizonsOut: Record<string, HorizonStats> = {}
  for (const h of horizons) {
    const vals = events.map(e => e.rets[String(h)]).filter((v): v is number => v != null && isFinite(v))
    if (vals.length === 0) {
      horizonsOut[String(h)] = { n: 0, mean: 0, median: 0, winRate: 0, p25: 0, p75: 0, best: 0, worst: 0 }
      continue
    }
    const sorted = [...vals].sort((a, b) => a - b)
    horizonsOut[String(h)] = {
      n: vals.length,
      mean: +mean(vals).toFixed(4),
      median: +quantile(sorted, 0.5).toFixed(4),
      winRate: +(vals.filter(v => v > 0).length / vals.length).toFixed(4),
      p25: +quantile(sorted, 0.25).toFixed(4),
      p75: +quantile(sorted, 0.75).toFixed(4),
      best: +sorted[sorted.length - 1].toFixed(4),
      worst: +sorted[0].toFixed(4),
    }
  }

  return { nEvents: events.length, events, horizons: horizonsOut }
}

// ==== 信号分级 ====

export type SignalStrength = 'strong' | 'moderate' | 'weak'

export function strengthOf(z: number): SignalStrength {
  if (z >= 2) return 'strong'
  if (z >= 1) return 'moderate'
  return 'weak'
}

export const STRENGTH_LABEL: Record<SignalStrength, string> = {
  strong: '强',
  moderate: '中',
  weak: '弱',
}

File: sync/analysis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""通用分析引擎：相关性 / 事件研究 / 统计工具。

这些函数原为 src/lib/analysis.ts 的 1:1 翻译（重计算从 Worker 迁移到 Python 同步层时），
用于保证两端数值一致。现前端已不再持有计算逻辑（改为直接读 analysis_results 的
预计算 JSONB），TS 版已随迁移完成删除，此模块即为唯一实现。

数值字段统一 round(value, 4) 处理后输出（对应原 TS 的 .toFixed(4)）。
"""
from math import log as _ln, sqrt, isfinite


# ── 统计基础 ──
def mean(nums):
    if not nums:
        return 0
    return sum(nums) / len(nums)


def std(nums):
    if len(nums) < 2:
        return 0
    m = mean(nums)
    return sqrt(sum((v - m) ** 2 for v in nums) / len(nums))


def percentile_rank(values, v):
    if len(values) < 2:
        return 50
    sorted_vals = sorted(values)
    below = sum(1 for x in sorted_vals if x < v)
    return round((below / (len(sorted_vals) - 1)) * 100, 1)


def z_score(values, v):
    m = mean(values)
    s = std(values)
    if s == 0:
        return 0
    return (v - m) / s


def quantile(sorted_vals, q):
    if not sorted_vals:
        return 0
    pos = (len(sorted_vals) - 1) * q
    lo = int(pos // 1)
    hi = int(-(-pos // 1))  # ceil
    if lo == hi:
        return sorted_vals[lo]
    return sorted_vals[lo] + (sorted_vals[hi] - sorted_vals[lo]) * (pos - lo)


# ── 收益率 ──
def log_returns(points):
    """相邻日 ln(p_i / p_{i-1})，首项跳过。points: [{date, value}, ...]"""
    out = []
    for i in range(1, len(points)):
        prev = points[i - 1]["value"]
        cur = points[i]["value"]
        if prev is not None and cur is not None and prev > 0 and cur > 0:
            out.append({"date": points[i]["date"], "value": _ln(cur / prev)})
    return out


# ── 按日期对齐两条序列（交集） ──
def align_by_date(a, b):
    """a, b: [{date, value}, ...] -> [{date, a, b}, ...]"""
    b_map = {p["date"]: p["value"] for p in b}
    out = []
    for p in a:
        bv = b_map.get(p["date"])
        if (
            bv is not None
            and isfinite(bv)
            and p["value"] is not None
            and isfinite(p["value"])
        ):
            out.append({"date": p["date"], "a": p["value"], "b": bv})
    return out


# ── 相关 ──
def corr(ax, bx):
    """Pearson 相关系数，去掉对齐后取最后 n 个，结果 +round 4 位"""
    n = min(len(ax), len(bx))
    if n < 3:
        return 0
    a = list(ax[-n:])
    b = list(bx[-n:])
    ma = mean(a)
    mb = mean(b)
    cov = sa2 = sb2 = 0
    for i in range(n):
        da = a[i] - ma
        db = b[i] - mb
        cov += da * db
        sa2 += da * da
        sb2 += db * db
    if sa2 == 0 or sb2 == 0:
        return 0
    return round(cov / sqrt(sa2 * sb2), 4)


def rolling_corr(a, b, window):
    """滑动窗口相关系数：每窗口计算一次 corr，输出每个右端点的 (date, value)"""
    aligned = align_by_date(a, b)
    out = []
    for i in range(window, len(aligned) + 1):
        seg = aligned[i - window:i]
        r = corr([pt["a"] for pt in seg], [pt["b"] for pt in seg])
        out.append({"date": seg[-1]["date"], "value": r})
    return out


# ── 事件研究 ──
def _forward_returns(price_dates, price_map, base_date, horizons):
    """以 base_date 之前最近一个交易日为基准，计算 N 个交易日后收益率（小数，+round 4）"""
    base_idx = -1
    for i, d in enumerate(price_dates):
        if d <= base_date:
            base_idx = i
        else:
            break
    if base_idx < 0:
        return {}
    base = price_map.get(price_dates[base_idx])
    if base is None or base == 0:
        return {}
    rets = {}
    for h in horizons:
        j = base_idx + h
        if j < len(price_dates):
            fwd = price_map.get(price_dates[j])
            rets[str(h)] = round((fwd / base) - 1, 4) if fwd is not None and fwd > 0 else None
        else:
            rets[str(h)] = None
    return rets


def event_study(prices, event_dates, horizons):
    """prices: [{date, value}], event_dates: ['YYYY-MM-DD', ...], horizons: [int, ...]

    返回 dict:
      {
        'nEvents': int,
        'events': [{date, rets: {horizon: r|null}}, ...],
        'horizons': {h: HorizonStats}
      }
    """
    price_dates = sorted(str(p["date"]) for p in prices)
    price_map = {str(p["date"]): p["value"] for p in prices}

    events = []
    for ev in event_dates:
        rets = _forward_returns(price_dates, price_map, str(ev), horizons)
        events.append({"date": str(ev), "rets": rets})

    horizons_out = {}
    for h in horizons:
        vals = [
            e["rets"][str(h)]
            for e in events
            if e["rets"].get(str(h)) is not None and isfinite(e["rets"][str(h)])
        ]
        if not vals:
            horizons_out[str(h)] = {
                "n": 0, "mean": 0, "median": 0, "winRate": 0,
                "p25": 0, "p75": 0, "best": 0, "worst": 0,
            }
            continue
        sorted_vals = sorted(vals)
        horizons_out[str(h)] = {
            "n": len(vals),
            "mean": round(mean(vals), 4),
            "median": round(quantile(sorted_vals, 0.5), 4),
            "winRate": round(sum(1 for v in vals if v > 0) / len(vals), 4),
            "p25": round(quantile(sorted_vals, 0.25), 4),
            "p75": round(quantile(sorted_vals, 0.75), 4),
            "best": round(sorted_vals[-1], 4),
            "worst": round(sorted_vals[0], 4),
        }

    return {"nEvents": len(events), "events": events, "horizons": horizons_out}